java.lang.OutOfMemoryError: Java heap space when OpenAPI is enabled and enums are used in a JAX-RS endpoint

As reported in https://github.com/quarkusio/quarkus/issues/4329, an OOME is thrown when the quarkus-smallrye-openapi extension is used and a JAX-RS endpoint has an enum as a query parameter. It works if smallrye-openapi 1.1.3 is used

@EricWittmann - I was able to reproduce it locally, and with a slightly different stack trace:
```
Caused by: io.quarkus.builder.BuildException: 
Build failure: Build failed due to errors
        [error]: Build step io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor#build threw an exception: java.lang.OutOfMemoryError: Java heap space
        at java.base/java.util.ArrayList.iterator(ArrayList.java:976)
        at org.jboss.jandex.CompositeIndex.getClassByName(CompositeIndex.java:189)
        at io.quarkus.arc.processor.BeanArchives$IndexWrapper.getClassByName(BeanArchives.java:101)
        at org.jboss.jandex.CompositeIndex.getClassByName(CompositeIndex.java:190)
        at io.smallrye.openapi.runtime.scanner.FilteredIndexView.getClassByName(FilteredIndexView.java:108)
        at org.jboss.jandex.CompositeIndex.getClassByName(CompositeIndex.java:190)
        at io.smallrye.openapi.runtime.scanner.dataobject.AugmentedIndexView.getClassByName(AugmentedIndexView.java:63)
        at io.smallrye.openapi.runtime.util.TypeUtil.isA(TypeUtil.java:170)
        at io.smallrye.openapi.runtime.scanner.dataobject.TypeProcessor.isA(TypeProcessor.java:280)
        at io.smallrye.openapi.runtime.scanner.dataobject.TypeProcessor.processType(TypeProcessor.java:136)
        at io.smallrye.openapi.runtime.scanner.dataobject.AnnotationTargetProcessor.readUnannotatedField(AnnotationTargetProcessor.java:184)
        at io.smallrye.openapi.runtime.scanner.dataobject.AnnotationTargetProcessor.processField(AnnotationTargetProcessor.java:134)
        at io.smallrye.openapi.runtime.scanner.dataobject.AnnotationTargetProcessor.process(AnnotationTargetProcessor.java:96)
        at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.depthFirstGraphSearch(OpenApiDataObjectScanner.java:265)
        at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:232)
        at io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:168)
        at io.smallrye.openapi.runtime.util.SchemaFactory.introspectClassToSchema(SchemaFactory.java:267)
        at io.smallrye.openapi.runtime.util.SchemaFactory.typeToSchema(SchemaFactory.java:233)
        at io.smallrye.openapi.runtime.scanner.ParameterProcessor.getParameters(ParameterProcessor.java:540)
        at io.smallrye.openapi.runtime.scanner.ParameterProcessor.process(ParameterProcessor.java:400)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsMethod(OpenApiAnnotationScanner.java:582)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.lambda$processJaxRsResourceClass$0(OpenApiAnnotationScanner.java:435)
        at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner$$Lambda$504/0x0000000840338c40.accept(Unknown Source)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
```
I used the Quarkus ```getting-started``` project along with this resource:
```java
@Path("/hello")
public class GreetingResource {

	public enum TestEnum {
		VAL1, VAL2, VAL3;
	}

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @Operation(operationId = "hello")
    public String hello(@QueryParam("val") TestEnum value) {
        return "hello";
    }
}
```
Even being less heavy handed and excluding only ```java.lang.Object``` resolves this and significantly speeds up the scanning. Without being more familiar with how the index is originally built, it does appear from that trace that there are multiple levels of indexes. I have not looked at Jandex to understand the memory impacts of that.

Additionally, we are not supporting enums in parameters or responses properly - they both end up being simple objects. Fixing that part of the scanning by properly generating string schemas with the list of enum values does fix the immediate issue, but I think stopping the hierarchy scan when a parent class is ```java.lang.Object``` is the way to go to prevent this more generally.

--- a/implementation/pom.xml
+++ b/implementation/pom.xml
@@ -29,16 +29,6 @@
 
     <name>SmallRye: MicroProfile OpenAPI Implementation</name>
 
-    <dependencyManagement>
-        <dependencies>
-            <dependency>
-                <groupId>com.fasterxml.jackson.core</groupId>
-                <artifactId>jackson-annotations</artifactId>
-                <scope>test</scope>
-            </dependency>
-        </dependencies>
-    </dependencyManagement>
-
     <dependencies>
         <!-- MP OpenAPI Spec -->
         <dependency>

--- a/implementation/src/main/java/io/smallrye/openapi/api/OpenApiConfigImpl.java
+++ b/implementation/src/main/java/io/smallrye/openapi/api/OpenApiConfigImpl.java
@@ -127,6 +127,7 @@ public class OpenApiConfigImpl implements OpenApiConfig {
         if (scanExcludePackages == null) {
             String packages = getConfig().getOptionalValue(OASConfig.SCAN_EXCLUDE_PACKAGES, String.class).orElse(null);
             scanExcludePackages = asCsvSet(packages);
+            scanExcludePackages.addAll(OpenApiConstants.NEVER_SCAN_PACKAGES);
         }
         return scanExcludePackages;
     }
@@ -139,6 +140,7 @@ public class OpenApiConfigImpl implements OpenApiConfig {
         if (scanExcludeClasses == null) {
             String classes = getConfig().getOptionalValue(OASConfig.SCAN_EXCLUDE_CLASSES, String.class).orElse(null);
             scanExcludeClasses = asCsvSet(classes);
+            scanExcludeClasses.addAll(OpenApiConstants.NEVER_SCAN_CLASSES);
         }
         return scanExcludeClasses;
     }

--- a/implementation/src/main/java/io/smallrye/openapi/api/OpenApiConstants.java
+++ b/implementation/src/main/java/io/smallrye/openapi/api/OpenApiConstants.java
@@ -80,6 +80,18 @@ public final class OpenApiConstants {
     public static final String SCHEMA_REFERENCES_ENABLE = "mp.openapi.extensions.schema-references.enable";
     public static final String CUSTOM_SCHEMA_REGISTRY_CLASS = "mp.openapi.extensions.custom-schema-registry.class";
 
+    /**
+     * Set of classes which should never be scanned, regardless of user configuration.
+     */
+    public static final Set<String> NEVER_SCAN_CLASSES = Collections
+            .unmodifiableSet(new HashSet<>(Arrays.asList()));
+
+    /**
+     * Set of packages which should never be scanned, regardless of user configuration.
+     */
+    public static final Set<String> NEVER_SCAN_PACKAGES = Collections
+            .unmodifiableSet(new HashSet<>(Arrays.asList("java.lang")));
+
     public static final String CLASS_SUFFIX = ".class";
     public static final String JAR_SUFFIX = ".jar";
     public static final String WEB_ARCHIVE_CLASS_PREFIX = "/WEB-INF/classes/";

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiAnnotationScanner.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiAnnotationScanner.java
@@ -164,7 +164,13 @@ public class OpenApiAnnotationScanner {
      */
     public OpenApiAnnotationScanner(OpenApiConfig config, IndexView index, List<AnnotationScannerExtension> extensions) {
         this.config = config;
-        this.index = index;
+
+        if (index instanceof FilteredIndexView) {
+            this.index = index;
+        } else {
+            this.index = new FilteredIndexView(index, config);
+        }
+
         this.extensions = extensions;
     }
 

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScanner.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScanner.java
@@ -191,11 +191,15 @@ public class OpenApiDataObjectScanner {
         LOG.debugv("Starting processing with root: {0}", rootClassType.name());
 
         // If top level item is simple
-        if (isTerminalType(rootClassType)) {
+        if (TypeUtil.isTerminalType(rootClassType)) {
             SchemaImpl simpleSchema = new SchemaImpl();
             simpleSchema.setType(rootClassTypeFormat.getSchemaType());
             simpleSchema.setFormat(rootClassTypeFormat.getFormat().format());
             return simpleSchema;
+        }
+
+        if (isA(rootClassType, ENUM_TYPE) && index.containsClass(rootClassType)) {
+            return SchemaFactory.enumToSchema(index, rootClassType);
         }
 
         // If top level item is not indexed
@@ -274,24 +278,6 @@ public class OpenApiDataObjectScanner {
 
     private boolean isA(Type testSubject, Type test) {
         return TypeUtil.isA(index, testSubject, test);
-    }
-
-    private boolean isTerminalType(Type type) {
-        if (type.kind() == Type.Kind.TYPE_VARIABLE ||
-                type.kind() == Type.Kind.WILDCARD_TYPE ||
-                type.kind() == Type.Kind.ARRAY) {
-            return false;
-        }
-
-        if (type.kind() == Type.Kind.PRIMITIVE ||
-                type.kind() == Type.Kind.VOID) {
-            return true;
-        }
-
-        TypeUtil.TypeWithFormat tf = TypeUtil.getTypeFormat(type);
-        // If is known type.
-        return tf.getSchemaType() != Schema.SchemaType.OBJECT &&
-                tf.getSchemaType() != Schema.SchemaType.ARRAY;
     }
 
     // Is Map, Collection, etc.

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/dataobject/TypeProcessor.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/dataobject/TypeProcessor.java
@@ -21,18 +21,19 @@ import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.ENUM_
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.MAP_TYPE;
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.OBJECT_TYPE;
 import static io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.STRING_TYPE;
+import static io.smallrye.openapi.runtime.util.TypeUtil.isTerminalType;
 
 import org.eclipse.microprofile.openapi.models.media.Schema;
 import org.jboss.jandex.AnnotationTarget;
 import org.jboss.jandex.ArrayType;
-import org.jboss.jandex.ClassInfo;
-import org.jboss.jandex.FieldInfo;
 import org.jboss.jandex.ParameterizedType;
 import org.jboss.jandex.Type;
 import org.jboss.logging.Logger;
 
 import io.smallrye.openapi.api.models.media.SchemaImpl;
+import io.smallrye.openapi.api.util.MergeUtil;
 import io.smallrye.openapi.runtime.scanner.SchemaRegistry;
+import io.smallrye.openapi.runtime.util.SchemaFactory;
 import io.smallrye.openapi.runtime.util.TypeUtil;
 
 /**
@@ -110,20 +111,18 @@ public class TypeProcessor {
         }
 
         if (isA(type, ENUM_TYPE) && index.containsClass(type)) {
-            readEnumeration(type, schema);
+            MergeUtil.mergeObjects(schema, SchemaFactory.enumToSchema(index, type));
             return STRING_TYPE;
         }
 
         if (type.kind() == Type.Kind.PARAMETERIZED_TYPE) {
-            // Parameterised type (e.g. Foo<A, B>)
-            //return readParamType(annotationTarget, pathEntry, schema, fieldType.asParameterizedType(), typeResolver);
+            // Parameterized type (e.g. Foo<A, B>)
             return readParameterizedType(type.asParameterizedType());
         }
 
         if (type.kind() == Type.Kind.TYPE_VARIABLE ||
                 type.kind() == Type.Kind.UNRESOLVED_TYPE_VARIABLE) {
             // Resolve type variable to real variable.
-            //return resolveTypeVariable(annotationTarget, typeResolver, schema, pathEntry, fieldType);
             return resolveTypeVariable(schema, type);
         }
 
@@ -151,6 +150,7 @@ public class TypeProcessor {
 
     private Type readParameterizedType(ParameterizedType pType) {
         LOG.debugv("Processing parameterized type {0}", pType);
+        Type typeRead = pType;
 
         // If it's a collection, we should treat it as an array.
         if (isA(pType, COLLECTION_TYPE)) { // TODO maybe also Iterable?
@@ -171,7 +171,7 @@ public class TypeProcessor {
 
             schema.items(arraySchema);
 
-            return ARRAY_TYPE_OBJECT; // Representing collection as JSON array
+            typeRead = ARRAY_TYPE_OBJECT; // Representing collection as JSON array
         } else if (isA(pType, MAP_TYPE)) {
             LOG.debugv("Processing Map. Will treat as an object.");
             schema.type(Schema.SchemaType.OBJECT);
@@ -189,12 +189,13 @@ public class TypeProcessor {
                 // Add properties schema to field schema.
                 schema.additionalPropertiesSchema(propsSchema);
             }
-            return OBJECT_TYPE;
-        } else {
+            typeRead = OBJECT_TYPE;
+        } else if (index.containsClass(type)) {
             // This type will be resolved later, if necessary.
             pushToStack(pType);
-            return pType;
         }
+
+        return typeRead;
     }
 
     private Schema resolveParameterizedType(Type valueType, Schema propsSchema) {
@@ -209,8 +210,7 @@ public class TypeProcessor {
         } else if (index.containsClass(valueType)) {
             if (isA(valueType, ENUM_TYPE)) {
                 LOG.debugv("Processing an enum {0}", valueType);
-                propsSchema.type(Schema.SchemaType.STRING);
-                readEnumeration(valueType, propsSchema);
+                propsSchema = SchemaFactory.enumToSchema(index, valueType);
             } else {
                 propsSchema.type(Schema.SchemaType.OBJECT);
                 pushToStack(valueType, propsSchema);
@@ -220,30 +220,6 @@ public class TypeProcessor {
         }
 
         return propsSchema;
-    }
-
-    /**
-     * Add each Enum constant name to the list of the given schema's
-     * enumeration list.
-     *
-     * @param enumType type containing Java Enum constants
-     * @param enumSchema the schema to which the constants' names will
-     *        be added
-     *
-     * @see java.lang.reflect.Field#isEnumConstant()
-     * @see java.lang.reflect.Modifier#ENUM
-     */
-    private void readEnumeration(Type enumType, Schema enumSchema) {
-        LOG.debugv("Processing an enum {0}", enumType);
-        final int ENUM = 0x00004000;
-        ClassInfo enumKlazz = index.getClass(enumType);
-
-        enumKlazz.fields()
-                .stream()
-                .filter(field -> (field.flags() & ENUM) != 0)
-                .map(FieldInfo::name)
-                .sorted() // Make the order determinate
-                .forEach(enumSchema::addEnumeration);
     }
 
     private Type resolveTypeVariable(Schema schema, Type fieldType) {
@@ -278,23 +254,5 @@ public class TypeProcessor {
 
     private boolean isA(Type testSubject, Type test) {
         return TypeUtil.isA(index, testSubject, test);
-    }
-
-    private boolean isTerminalType(Type type) {
-        if (type.kind() == Type.Kind.TYPE_VARIABLE ||
-                type.kind() == Type.Kind.WILDCARD_TYPE ||
-                type.kind() == Type.Kind.ARRAY) {
-            return false;
-        }
-
-        if (type.kind() == Type.Kind.PRIMITIVE ||
-                type.kind() == Type.Kind.VOID) {
-            return true;
-        }
-
-        TypeUtil.TypeWithFormat tf = TypeUtil.getTypeFormat(type);
-        // If is known type.
-        return tf.getSchemaType() != Schema.SchemaType.OBJECT &&
-                tf.getSchemaType() != Schema.SchemaType.ARRAY;
     }
 }

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/dataobject/TypeResolver.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/dataobject/TypeResolver.java
@@ -45,6 +45,7 @@ import org.jboss.jandex.TypeVariable;
 import org.jboss.logging.Logger;
 
 import io.smallrye.openapi.api.OpenApiConstants;
+import io.smallrye.openapi.runtime.util.JandexUtil;
 import io.smallrye.openapi.runtime.util.TypeUtil;
 
 /**
@@ -329,7 +330,7 @@ public class TypeResolver {
     }
 
     public static Map<String, TypeResolver> getAllFields(AugmentedIndexView index, Type leaf, ClassInfo leafKlazz) {
-        Map<ClassInfo, Type> chain = inheritanceChain(index, leafKlazz, leaf);
+        Map<ClassInfo, Type> chain = JandexUtil.inheritanceChain(index, leafKlazz, leaf);
         Map<String, TypeResolver> properties = new LinkedHashMap<>();
         Deque<Map<String, Type>> stack = new ArrayDeque<>();
 
@@ -370,28 +371,6 @@ public class TypeResolver {
 
     private static boolean acceptField(FieldInfo field) {
         return !Modifier.isStatic(field.flags());
-    }
-
-    /**
-     * Builds an insertion-order map of a class's inheritance chain, starting
-     * with the klazz argument.
-     *
-     * @param index index for superclass retrieval
-     * @param klazz the class to retrieve inheritance
-     * @param type type of the klazz
-     * @return map of a class's inheritance chain/ancestry
-     */
-    private static Map<ClassInfo, Type> inheritanceChain(AugmentedIndexView index,
-            ClassInfo klazz,
-            Type type) {
-
-        Map<ClassInfo, Type> chain = new LinkedHashMap<>();
-
-        do {
-            chain.put(klazz, type);
-        } while ((type = klazz.superClassType()) != null && (klazz = index.getClass(type)) != null);
-
-        return chain;
     }
 
     /**

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/util/JandexUtil.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/util/JandexUtil.java
@@ -23,8 +23,11 @@ import java.math.BigDecimal;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 import org.eclipse.microprofile.openapi.models.parameters.Parameter;
 import org.jboss.jandex.AnnotationInstance;
@@ -339,31 +342,23 @@ public class JandexUtil {
     }
 
     /**
-     * Use the jandex index to find all jax-rs resource classes. This is done by searching for
+     * Use the Jandex index to find all jax-rs resource classes. This is done by searching for
      * all Class-level @Path annotations.
      * 
      * @param index IndexView
      * @return Collection of ClassInfo's
      */
     public static Collection<ClassInfo> getJaxRsResourceClasses(IndexView index) {
-        Collection<ClassInfo> resourceClasses = new ArrayList<>();
-        Collection<AnnotationInstance> pathAnnotations = index.getAnnotations(OpenApiConstants.DOTNAME_PATH);
-        for (AnnotationInstance pathAnno : pathAnnotations) {
-            AnnotationTarget annotationTarget = pathAnno.target();
-            if (annotationTarget.kind() == AnnotationTarget.Kind.CLASS) {
-                ClassInfo classInfo = (ClassInfo) annotationTarget;
-                if (Modifier.isInterface(classInfo.flags())) {
-                    if (index.getAllKnownImplementors(classInfo.name())
-                            .stream()
-                            .anyMatch(info -> !Modifier.isAbstract(info.flags()))) {
-                        resourceClasses.add(annotationTarget.asClass());
-                    }
-                } else {
-                    resourceClasses.add(annotationTarget.asClass());
-                }
-            }
-        }
-        return resourceClasses;
+        return index.getAnnotations(OpenApiConstants.DOTNAME_PATH)
+                .stream()
+                .map(AnnotationInstance::target)
+                .filter(target -> target.kind() == AnnotationTarget.Kind.CLASS)
+                .map(AnnotationTarget::asClass)
+                .filter(classInfo -> !Modifier.isInterface(classInfo.flags()) ||
+                        index.getAllKnownImplementors(classInfo.name()).stream()
+                                .anyMatch(info -> !Modifier.isAbstract(info.flags())))
+                .distinct() // CompositeIndex instances may return duplicates
+                .collect(Collectors.toList());
     }
 
     /**
@@ -561,6 +556,26 @@ public class JandexUtil {
                 OpenApiConstants.PROP_TYPE, org.eclipse.microprofile.openapi.models.media.Schema.SchemaType.class);
         String implementation = JandexUtil.stringValue(annotation, OpenApiConstants.PROP_IMPLEMENTATION);
         return (type == org.eclipse.microprofile.openapi.models.media.Schema.SchemaType.ARRAY && implementation != null);
+    }
+
+    /**
+     * Builds an insertion-order map of a class's inheritance chain, starting
+     * with the klazz argument.
+     *
+     * @param index index for superclass retrieval
+     * @param klazz the class to retrieve inheritance
+     * @param type type of the klazz
+     * @return map of a class's inheritance chain/ancestry
+     */
+    public static Map<ClassInfo, Type> inheritanceChain(IndexView index, ClassInfo klazz, Type type) {
+        Map<ClassInfo, Type> chain = new LinkedHashMap<>();
+
+        do {
+            chain.put(klazz, type);
+        } while ((type = klazz.superClassType()) != null &&
+                (klazz = index.getClassByName(TypeUtil.getName(type))) != null);
+
+        return chain;
     }
 
     /**

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/util/SchemaFactory.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/util/SchemaFactory.java
@@ -12,7 +12,9 @@ import org.eclipse.microprofile.openapi.models.media.Schema;
 import org.eclipse.microprofile.openapi.models.media.Schema.SchemaType;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationValue;
+import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.ClassType;
+import org.jboss.jandex.FieldInfo;
 import org.jboss.jandex.IndexView;
 import org.jboss.jandex.ParameterizedType;
 import org.jboss.jandex.Type;
@@ -217,7 +219,7 @@ public class SchemaFactory {
     }
 
     /**
-     * Converts a jandex type to a {@link Schema} model.
+     * Converts a Jandex type to a {@link Schema} model.
      * 
      * @param index the index of classes being scanned
      * @param type the implementation type of the item to scan
@@ -244,6 +246,35 @@ public class SchemaFactory {
         }
 
         return schema;
+    }
+
+    /**
+     * Convert a Jandex enum class type to a {@link Schema} model.
+     * 
+     * Adds each enum constant name to the list of the given schema's
+     * enumeration list. The given type must be found in the index.
+     *
+     * @param index Jandex index containing the ClassInfo for the given enum type
+     * @param enumType type containing Java Enum constants
+     *
+     * @see java.lang.reflect.Field#isEnumConstant()
+     * @see java.lang.reflect.Modifier#ENUM
+     */
+    public static Schema enumToSchema(IndexView index, Type enumType) {
+        LOG.debugv("Processing an enum {0}", enumType);
+        final int ENUM = 0x00004000;
+        ClassInfo enumKlazz = index.getClassByName(TypeUtil.getName(enumType));
+        Schema enumSchema = new SchemaImpl();
+        enumSchema.setType(SchemaType.STRING);
+
+        enumKlazz.fields()
+                .stream()
+                .filter(field -> (field.flags() & ENUM) != 0)
+                .map(FieldInfo::name)
+                .sorted() // Make the order determinate
+                .forEach(enumSchema::addEnumeration);
+
+        return enumSchema;
     }
 
     /**

--- a/implementation/src/main/java/io/smallrye/openapi/runtime/util/TypeUtil.java
+++ b/implementation/src/main/java/io/smallrye/openapi/runtime/util/TypeUtil.java
@@ -27,6 +27,7 @@ import java.util.stream.Collectors;
 
 import javax.validation.constraints.NotNull;
 
+import org.eclipse.microprofile.openapi.models.media.Schema;
 import org.eclipse.microprofile.openapi.models.media.Schema.SchemaType;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
@@ -201,6 +202,24 @@ public class TypeUtil {
             superKlazzType = superKlazz.superClassType();
         }
         return false;
+    }
+
+    public static boolean isTerminalType(Type type) {
+        if (type.kind() == Type.Kind.TYPE_VARIABLE ||
+                type.kind() == Type.Kind.WILDCARD_TYPE ||
+                type.kind() == Type.Kind.ARRAY) {
+            return false;
+        }
+
+        if (type.kind() == Type.Kind.PRIMITIVE ||
+                type.kind() == Type.Kind.VOID) {
+            return true;
+        }
+
+        TypeUtil.TypeWithFormat tf = TypeUtil.getTypeFormat(type);
+        // If is known type.
+        return tf.getSchemaType() != Schema.SchemaType.OBJECT &&
+                tf.getSchemaType() != Schema.SchemaType.ARRAY;
     }
 
     public static DotName getName(Type type) {

--- a/implementation/src/test/java/io/smallrye/openapi/runtime/scanner/ParameterScanTests.java
+++ b/implementation/src/test/java/io/smallrye/openapi/runtime/scanner/ParameterScanTests.java
@@ -161,6 +161,13 @@ public class ParameterScanTests extends IndexScannerTestBase {
                 Widget.class);
     }
 
+    @Test
+    public void testEnumQueryParam() throws IOException, JSONException {
+        test("params.enum-form-param.json",
+                EnumQueryParamTestResource.class,
+                EnumQueryParamTestResource.TestEnum.class);
+    }
+
     /***************** Test models and resources below. ***********************/
 
     public static class Widget {
@@ -440,6 +447,23 @@ public class ParameterScanTests extends IndexScannerTestBase {
         public CompletionStage<Widget> upd(@MultipartForm Bean form,
                 @FormParam("f3") @DefaultValue("3") int formField3,
                 @org.jboss.resteasy.annotations.jaxrs.FormParam @NotNull String formField4) {
+            return null;
+        }
+    }
+
+    @Path("/enum/formparam")
+    static class EnumQueryParamTestResource {
+        static enum TestEnum {
+            VAL1,
+            VAL2,
+            VAL3;
+        }
+
+        @SuppressWarnings("unused")
+        @POST
+        @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+        @Produces(MediaType.TEXT_PLAIN)
+        public TestEnum postData(@QueryParam("val") TestEnum value) {
             return null;
         }
     }

--- a/implementation/src/test/resources/io/smallrye/openapi/runtime/scanner/params.enum-form-param.json
+++ b/implementation/src/test/resources/io/smallrye/openapi/runtime/scanner/params.enum-form-param.json
@@ -1,0 +1,38 @@
+{
+  "openapi": "3.0.1",
+  "paths": {
+    "/enum/formparam": {
+      "post": {
+        "parameters": [
+          {
+          "name": "val",
+          "in": "query",
+          "schema": {
+            "$ref": "#/components/schemas/TestEnum"
+          }
+        }
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "text/plain": {
+                "schema": {
+                  "$ref": "#/components/schemas/TestEnum"
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  },
+  "components": {
+    "schemas": {
+      "TestEnum": {
+        "type": "string",
+        "enum": [ "VAL1", "VAL2", "VAL3" ]
+      }
+    }
+  }
+}

